module/grpcserver/server_builder.go

```diff
@@ -0,0 +1,87 @@
+package grpcserver
+
+import (
+	"github.com/rs/zerolog"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	grpc_prometheus "github.com/grpc-ecosystem/go-grpc-prometheus"
+
+	"github.com/onflow/flow-go/engine/common/rpc"
+)
+
+type Option func(*GrpcServerBuilder)
+
+// WithTransportCredentials sets the transport credentials parameters for a grpc server builder.
+func WithTransportCredentials(transportCredentials credentials.TransportCredentials) Option {
+	return func(c *GrpcServerBuilder) {
+		c.transportCredentials = transportCredentials
+	}
+}
+
+// GrpcServerBuilder created for separating the creation and starting GrpcServer,
+// cause services need to be registered before the server starts.
+type GrpcServerBuilder struct {
+	log            zerolog.Logger
+	gRPCListenAddr string
+	server         *grpc.Server
+
+	transportCredentials credentials.TransportCredentials // the GRPC credentials
+}
+
+// NewGrpcServerBuilder helps to build a new grpc server.
+func NewGrpcServerBuilder(log zerolog.Logger,
+	gRPCListenAddr string,
+	maxMsgSize uint,
+	rpcMetricsEnabled bool,
+	apiRateLimits map[string]int, // the api rate limit (max calls per second) for each of the Access API e.g. Ping->100, GetTransaction->300
+	apiBurstLimits map[string]int, // the api burst limit (max calls at the same time) for each of the Access API e.g. Ping->50, GetTransaction->10
+	opts ...Option,
+) *GrpcServerBuilder {
+	log = log.With().Str("component", "grpc_server").Logger()
```

Comment:
at some point after applying the opts, can you add another log field indicating if this is the secure or plaintext endpoint.